pmtiles/bucket.go

```diff
@@ -1,65 +1,157 @@
 package pmtiles
 
 import (
+	"bytes"
 	"context"
+	"crypto/md5"
+	"encoding/hex"
+	"errors"
 	"fmt"
-	"gocloud.dev/blob"
 	"io"
 	"net/http"
 	"net/url"
 	"os"
 	"path"
 	"path/filepath"
 	"strings"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"gocloud.dev/blob"
 )
 
 // Bucket is an abstration over a gocloud or plain HTTP bucket.
 type Bucket interface {
 	Close() error
 	NewRangeReader(ctx context.Context, key string, offset int64, length int64) (io.ReadCloser, error)
+	NewRangeReaderEtag(ctx context.Context, key string, offset int64, length int64, etag string) (io.ReadCloser, string, error)
+}
+
+type RefreshRequiredError struct {
+	StatusCode int
+}
+
+func (m *RefreshRequiredError) Error() string {
+	return fmt.Sprintf("HTTP error indicates file has changed: %d", m.StatusCode)
+}
+
+type mockBucket struct {
+	items map[string][]byte
+}
+
+func (m mockBucket) Close() error {
+	return nil
+}
+
+func (m mockBucket) NewRangeReader(ctx context.Context, key string, offset int64, length int64) (io.ReadCloser, error) {
+	body, _, err := m.NewRangeReaderEtag(ctx, key, offset, length, "")
+	return body, err
+
+}
+func (m mockBucket) NewRangeReaderEtag(ctx context.Context, key string, offset int64, length int64, etag string) (io.ReadCloser, string, error) {
+	bs, ok := m.items[key]
+	if !ok {
+		return nil, "", fmt.Errorf("Not found %s", key)
+	}
+
+	hash := md5.Sum(bs)
+	resultEtag := hex.EncodeToString(hash[:])
+	if len(etag) > 0 && resultEtag != etag {
+		return nil, "", &RefreshRequiredError{}
+	}
+	if offset+length > int64(len(bs)) {
+		return nil, "", &RefreshRequiredError{416}
+	}
+
+	return io.NopCloser(bytes.NewReader(bs[offset:(offset + length)])), resultEtag, nil
+}
+
+type HttpClient interface {
+	Do(req *http.Request) (*http.Response, error)
 }
 
 type HTTPBucket struct {
 	baseURL string
+	client  HttpClient
 }
 
-func (b HTTPBucket) NewRangeReader(_ context.Context, key string, offset, length int64) (io.ReadCloser, error) {
+func (b HTTPBucket) NewRangeReader(ctx context.Context, key string, offset, length int64) (io.ReadCloser, error) {
+	body, _, err := b.NewRangeReaderEtag(ctx, key, offset, length, "")
+	return body, err
+}
+
+func (b HTTPBucket) NewRangeReaderEtag(ctx context.Context, key string, offset, length int64, etag string) (io.ReadCloser, string, error) {
 	reqURL := b.baseURL + "/" + key
 
-	req, err := http.NewRequest("GET", reqURL, nil)
+	req, err := http.NewRequestWithContext(ctx, "GET", reqURL, nil)
```

Comment:
changed this to pass through the context, let me know if there was a reason not to